### PR TITLE
FIR checker: add diagnostics for backing fields

### PR DESCRIPTION
**Pull request by jsjeon**
This PR revisits checker utils for property initializers, where some of candidate diagnostics for backing fields were not enabled. The main reason was due to the lack of the counter part of `BindingContext.BACKING_FIELD_REQUIRED` on FIR side. It turns out that that doesn't require non-trivial computation; it just needs a correct modality of property accessors, which is missed before. 1st commit address that first, and 2nd commit adds diagnostics:
* BACKING_FIELD_IN_INTERFACE
* PROPERTY_INITIALIZER_NO_BACKING_FIELD
* EXTENSION_PROPERTY_WITH_BACKING_FIELD

### CODE REVIEW

File: compiler/testData/diagnostics/testsWithStdLib/contracts/dsl/errors/contractCallSites.1.3.fir.kt [hunk 1]
```diff
@@ -56,7 +56,7 @@ val topLevelAnonymousFunction = fun (x: Boolean) {
     contract { returns() implies (x) }
 }
 
-var topLevelPropertyAccessors: Int? = 42
+var topLevelPropertyAccessors: Int? = <!PROPERTY_INITIALIZER_NO_BACKING_FIELD!>42<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/testsWithStdLib/contracts/dsl/errors/contractCallSites.1.3.fir.kt:59]: This and below are false alarms, but mainly due to the unresolved backing field. Filed https://youtrack.jetbrains.com/issue/KT-44681

File: compiler/testData/diagnostics/tests/scopes/initializerScopeOfExtensionProperty.fir.kt [hunk 2]
```diff
@@ -24,8 +24,8 @@ class C {
     inner class D {}
 }
 
-val C.foo : C.D = <!UNRESOLVED_REFERENCE!>D<!>()
+val C.foo : C.D = <!EXTENSION_PROPERTY_WITH_BACKING_FIELD!><!UNRESOLVED_REFERENCE!>D<!>()<!>
 
-val C.bar : C.D = C().D()
+val C.bar : C.D = C().<!EXTENSION_PROPERTY_WITH_BACKING_FIELD!>D()<!>
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/scopes/initializerScopeOfExtensionProperty.fir.kt:29]: This should be reported on `C().D()`, not `D()`. I explained why, and suggested a couple way to fix: https://youtrack.jetbrains.com/issue/KT-42591 PTAL that issue and leave comments about what option you prefer, there or here.

File: compiler/testData/diagnostics/testsWithStdLib/contracts/dsl/errors/contractCallSites.1.3.fir.kt (hunk 1 above)
Comment (jsjeon) [reply compiler/testData/diagnostics/testsWithStdLib/contracts/dsl/errors/contractCallSites.1.3.fir.kt]: 3rd commit adds a local scope for backing fields before resolving property accessors, and now they're properly resolved. Also, these false alarms are gone too.

Comment (jsjeon) [reply compiler/testData/diagnostics/testsWithStdLib/contracts/dsl/errors/contractCallSites.1.3.fir.kt]: According to @demiurg906, this is by design: https://youtrack.jetbrains.com/issue/KT-44681#focus=Comments-27-4674827.0-0 I'll revert the 3rd commit.

**Comment by mglukhikh** [comment]
I consider all cases with `abstract get` or `abstract set` as very special, because `abstract` modifier is anyway forbidden on accessors. Don't see much reason to store accessors modality. Will do necessary changes myself.

**Comment by mglukhikh** [comment]
Also, I'm not sure that separate `hasBackingField` and `backingFieldRequired` are needed for something. Will check.

**Comment by jsjeon** [comment]
Alright, just in case, I rebased the branch on top of recent changes about generator-based diagnostic list.

**Comment by mglukhikh** [comment]
Merged manually. Resulting commit: https://github.com/JetBrains/kotlin/commit/bd37badf297e82a97b7cd05ffa49ecf0ad2ec203. Please pay attention to the new diagnostic DSL by Ilya Kirillov.

**Comment by mglukhikh** [comment]
> I check that as well, but I can tell what backend IR expects properties to have a backing field v.s. what FE requires properties to have a backing field are different, and thus not quite interchangeable.

Yes, there is a questionable place with delegated property `val x: Type by something { ... }`. In different places it may be considered as having or not having backing field, so delegate presence should be checked separately, if needed.

**Comment by jsjeon** [comment]
> Please pay attention to the new diagnostic DSL by Ilya Kirillov.

Yeah, Dmitriy let me know what task I need to run afterwards. Thanks for the fix.

>> I check that as well, but I can tell what backend IR expects properties to have a backing field v.s. what FE requires properties to have a backing field are different, and thus not quite interchangeable.
>
> Yes, there is a questionable place with delegated property val x: Type by something { ... }. In different places it may be considered as having or not having backing field, so delegate presence should be checked separately, if needed.

Oh, but you managed to merge `hasBackingField` somehow, nice!

File: compiler/testData/diagnostics/tests/scopes/initializerScopeOfExtensionProperty.fir.kt (hunk 2 above)
Comment (jsjeon) [reply compiler/testData/diagnostics/tests/scopes/initializerScopeOfExtensionProperty.fir.kt:29]: @mglukhikh Could you take a look at that issue about wrong source of qualified access expression?

**Comment by jsjeon** [comment]
> > Please pay attention to the new diagnostic DSL by Ilya Kirillov.
> 
> Yeah, Dmitriy let me know what task I need to run afterwards. Thanks for the fix.

cc @darthorimar 

FIR-IDE codegen I missed in this PR will be added at https://github.com/JetBrains/kotlin/pull/4101/commits/1b138c5bc0bb35420d2dcfa2ac9292d6e57d9b9a

File: compiler/testData/diagnostics/tests/scopes/initializerScopeOfExtensionProperty.fir.kt (hunk 2 above)
Comment (mglukhikh) [reply compiler/testData/diagnostics/tests/scopes/initializerScopeOfExtensionProperty.fir.kt:29]: Yes, I'm planning to look at that next week.

**Pull request closed without merging**